sdks/python/apache_beam/typehints/trivial_inference.py

```diff
@@ -65,18 +65,27 @@ def instance_to_type(o):
   elif t == tuple:
     return typehints.Tuple[[instance_to_type(item) for item in o]]
   elif t == list:
-    return typehints.List[
-        typehints.Union[[instance_to_type(item) for item in o]]
-    ]
+    if len(o):
+      return typehints.List[
+          typehints.Union[[instance_to_type(item) for item in o]]
+      ]
+    else:
+      return typehints.List[typehints.Any]
```

Comment:
I'm new to reviewing this - is `typehints.Any` equivalent to gradual typing "dynamic" type? Or it is the top type? (aka you can view any other value as Any but not vice versa).

Is there a separate type for a mutable list vs immutable list? That would affect whether this is a contravariant or invariant situation.

Reply:
I think it's equivalent. It's defined here: https://www.python.org/dev/peps/pep-0483/#summary-of-gradual-typing (look at the 3 bullet points).

Tuple is an immutable list. Its type hint has 2 forms:
- Tuple[type1[, type2[, ..]]], i.e., you specify the type for every index
- Tuple[type, ...], i.e., a homogenous list (note that `...` parses in Python as Ellipsis)
 
Could explain the contravariant vs invariant situation?

Reply:
ping: was your comment implying that I have a bug here or missing an edge case?